subprojects/resources/src/main/java/org/gradle/internal/resource/ExternalResourceName.java

```diff
@@ -52,29 +52,29 @@ private ExternalResourceName(String encodedRoot, String path) {
 
     public ExternalResourceName(URI parent, String path) {
         if (parent.getPath() == null) {
-            throw new IllegalArgumentException(String.format("Cannot create resource name from non-hierarchical URI '%s'.", parent.toString()));
+            throw new IllegalArgumentException(String.format("Cannot create resource name from non-hierarchical URI '%s'.", parent));
         }
-        String newPath;
+        this.encodedRoot = encodeRoot(parent);
+        this.path = combine(parent, path);
+    }
```

Comment:
Nit: maybe add a `private` constructor to set the fields, and let the other constructors call it instead of duplicating the setting logic multiple times?

Reply:
Thanks for the suggestion. Done in 0dc1368ef57549d04c8203c0070111345eeeb641